app/renderer/util.js

```diff
@@ -121,6 +137,119 @@ export function getOptimalXPath (doc, domNode, uniqueAttributes = ['id']) {
   }
 }
 
+/**
+ * Get an optimal Class Chain for a DOMNode based on the getOptimalXPath method
+ *
+ * @param {DOMDocument} doc
+ * @param {DOMNode} domNode
+ * @param {Array<String>} uniqueAttributes Attributes we know are unique
+ * @returns {string}
+ */
+function getOptimalClassChain (doc, domNode, uniqueAttributes) {
+  try {
+    // BASE CASE #1: If this isn't an element, we're above the root, or this is `XCUIElementTypeApplication`,
+    // which is not an official XCUITest element, return empty string
+    if (!domNode.tagName || domNode.nodeType !== 1 || domNode.tagName === 'XCUIElementTypeApplication') {
+      return '';
+    }
+
+    // BASE CASE #2: If this node has a unique class chain based on attributes then return it
+    for (let attrName of uniqueAttributes) {
+      const attrValue = domNode.getAttribute(attrName);
+      if (attrValue) {
+        let xpath = `//${domNode.tagName || '*'}[@${attrName}="${attrValue}"]`;
+        let classChain = `/${domNode.tagName || '*'}[\`${attrName} == "${attrValue}"\`]`;
+        let othersWithAttr;
+
+        // If the XPath does not parse, move to the next unique attribute
+        try {
+          othersWithAttr = XPath.select(xpath, doc);
+        } catch (ign) {
+          continue;
+        }
+
+        // If the attribute isn't actually unique, get it's index too
+        if (othersWithAttr.length > 1) {
+          let index = othersWithAttr.indexOf(domNode);
+          classChain = `${classChain}[${index + 1}]`;
+        }
+        return classChain;
+      }
+    }
+
+    // Get the relative xpath of this node using tagName
+    let classChain = `/${domNode.tagName}`;
+
+    // If this node has siblings of the same tagName, get the index of this node
+    if (domNode.parentNode) {
+      // Get the siblings
+      const childNodes = Array.prototype.slice.call(domNode.parentNode.childNodes, 0).filter((childNode) => (
+        childNode.nodeType === 1 && childNode.tagName === domNode.tagName
+      ));
+
+      // If there's more than one sibling, append the index
+      if (childNodes.length > 1) {
+        let index = childNodes.indexOf(domNode);
+        classChain += `[${index + 1}]`;
+      }
+    }
+
+    // Make a recursive call to this nodes parents and prepend it to this xpath
+    return getOptimalClassChain(doc, domNode.parentNode, uniqueAttributes) + classChain;
+  } catch (ign) {
+    // If there's an unexpected exception, abort and don't get an XPath
+    return null;
```

Comment:
I assume it makes sense to write the exception into the debug log. Otherwise it would be very hard to catch it

Reply:
This is also copied form the `getOptimalXPath` methods which also returned `null`, to keep it aligned

Reply:
the returned value is fine. I am concerned about logging the exception stack trace rather than silencing it completely